In [1]:
%matplotlib inline
import numpy as np
import scipy as sp
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [2]:
brooklyn=pd.read_csv('brooklyn.csv', low_memory=False)
queens=pd.read_csv('queens.csv', low_memory=False)
brooklyn_homes=pd.read_csv('brooklyn_homes.csv')
queens_homes=pd.read_csv('queens_homes.csv')
queens_low_val=pd.read_csv('queens_low_val.csv')
brooklyn_low_val=pd.read_csv('brooklyn_low_val.csv')
queens=queens.dropna()
brooklyn=brooklyn.dropna()
queens=queens.replace(' -   ','')
queens=queens.replace(' -   ','')

## Logistic Regression

In [3]:
queens.columns

Index(['Unnamed: 0', 'Unnamed: 0.1', 'address', 'apartment_number', 'block',
       'borough', 'building_class_at_present',
       'building_class_at_time_of_sale', 'building_class_category',
       'commercial_units', 'commerical_units', 'ease-ment',
       'gross_square_feet', 'gross_square_feet_c', 'land_square_feet', 'lot',
       'month', 'neighborhood', 'residential_units', 'sale_date', 'sale_price',
       'sale_price_clean', 'tax_class_at_present', 'tax_class_at_time_of_sale',
       'total_units', 'year', 'year_built', 'zip_code'],
      dtype='object')

In [4]:
Xqlr, Xqtestlr, yqlr, yqtestlr=train_test_split(queens[['sale_price_clean','year', 'month', 'gross_square_feet']].values, queens['neighborhood'].values, random_state=5)

In [5]:
clf=LogisticRegression()
clf.fit(Xqlr, yqlr)
print('Accuracy Score For Queens:', accuracy_score(clf.predict(Xqtestlr), yqtestlr))

Accuracy Score For Queens: 0.08618556701030927


This classification model is only 8.61% accurate in predicting the neighborhood in which a sale took place.

In [6]:
Xqlr1, Xqtestlr1, yqlr1, yqtestlr1=train_test_split(queens[['sale_price_clean','year', 'month', 'gross_square_feet']].values, queens['neighborhood'].values, random_state=5)

In [7]:
clf=LogisticRegression()
clf.fit(Xqlr1, yqlr1)
print('Accuracy Score For Queens:', accuracy_score(clf.predict(Xqtestlr1), yqtestlr1))

Accuracy Score For Queens: 0.08618556701030927


In [54]:
X1=pd.DataFrame(pd.get_dummies(queens[['building_class_category', 'tax_class_at_present', 'year', 'month', 'zip_code', 'building_class_at_time_of_sale', 'block', 'lot']]))

In [55]:
Xqlrd1, Xqtestlrd1, yqlrd1, yqtestlrd1=train_test_split(X1.values, queens['neighborhood'].values, random_state=5)

In [56]:
clf.fit(Xqlrd1, yqlrd1)
print('Accuracy Score Using Dummies:', accuracy_score(clf.predict(Xqtestlrd1), yqtestlrd1))

Accuracy Score Using Dummies: 0.26969072164948454


In [11]:
Xqhlr, Xqhtestlr, yqhlr, yqhtestlr=train_test_split(queens_homes[['sale_price_clean','year', 'gross_square_feet', 'month']].values, queens_homes['neighborhood'].values, random_state=5)

In [12]:
clf.fit(Xqhlr, yqhlr)
print('Accuracy Score For Queens Homes:', accuracy_score(clf.predict(Xqhtestlr), yqhtestlr))

Accuracy Score For Queens Homes: 0.1136004217185029


The model is marginally better at 11.36% correct for homes sold in Queens

In [13]:
Xblr, Xbtestlr, yblr, ybtestlr=train_test_split(brooklyn[['sale_price_clean','year', 'gross_square_feet', 'month']].values, brooklyn['neighborhood'].values, random_state=5)

In [14]:
clf.fit(Xblr, yblr)
print('Accuracy Score for Brooklyn:', accuracy_score(clf.predict(Xbtestlr), ybtestlr))

Accuracy Score for Brooklyn: 0.10139979515192898


The accuracy score for Brooklyn is marginally worse at 10.14% but that is still not acceptable.  Perhaps a model to predict the class of building sold could be more accurate

In [15]:
Xbhlr, Xbhtestlr, ybhlr, ybhtestlr=train_test_split(brooklyn_homes[['sale_price_clean','year', 'month', 'gross_square_feet']].values, brooklyn_homes['neighborhood'].values, random_state=5)

In [16]:
clf.fit(Xbhlr, ybhlr)
print('Accuracy Score For Brooklyn Homes:', accuracy_score(clf.predict(Xbhtestlr), ybhtestlr))

Accuracy Score For Brooklyn Homes: 0.11897810218978103


Again, the model is better for homes than the whole borough, but 11.898% is not acceptable.

In [17]:
Xqlr2, Xqtestlr2, yqlr2, yqtestlr2=train_test_split(queens[['sale_price_clean','year', 'month', 'gross_square_feet']].values, queens['building_class_category'].values, random_state=5)

In [18]:
clf.fit(Xqlr2, yqlr2)
print('Accuracy Score for Queens Building Category:', accuracy_score(clf.predict(Xqtestlr), yqtestlr2))

Accuracy Score for Queens Building Category: 0.5584879725085911


Although the score is much improved from the Neighborhood model (55.84% correct), it is still almost a 50-50 chance of getting the building category correct.

In [19]:
Xblr2, Xbtestlr2, yblr2, ybtestlr2=train_test_split(brooklyn[['sale_price_clean','year', 'month', 'gross_square_feet']].values, brooklyn['building_class_category'].values, random_state=5)

In [20]:
clf.fit(Xblr2, yblr2)
print('Accuracy Score for Brooklyn Building Category:', accuracy_score(clf.predict(Xbtestlr), ybtestlr2))

Accuracy Score for Brooklyn Building Category: 0.23386821440764766


The accuracy score for Brooklyn shows that the model is not as good at predicting building categories in Brooklyn, as it is in Queens.  23.39% is not a good accuracy score but it is better than 11.898% accuracy score for predicting Neighborhoods in Brooklyn.

In [21]:
Xqlr3, Xqtestlr3, yqlr3, yqtestlr3=train_test_split(queens[['sale_price_clean','gross_square_feet', 'month']].values, queens['year'].values, random_state=5)

In [22]:
clf.fit(Xqlr3, yqlr3)
print('Accuracy Score for Year of Sale In Queens:', accuracy_score(clf.predict(Xqtestlr3), yqtestlr3))

Accuracy Score for Year of Sale In Queens: 0.7499656357388316


In [23]:
Xblr3, Xbtestlr3, yblr3, ybtestlr3=train_test_split(brooklyn[['sale_price_clean','gross_square_feet_c', 'month']].values, brooklyn['year'].values, random_state=5)

In [24]:
clf.fit(Xblr3, yblr3)
print('Accuracy Score for Accuracy Score for Year of Sale In Brooklyn:', accuracy_score(clf.predict(Xbtestlr3), ybtestlr3))

Accuracy Score for Accuracy Score for Year of Sale In Brooklyn: 0.7606691703653123


The model is much better at predicting the year a sale took place in.  The year of a sale was correctly predicted at around 75% for both boroughs.

## Linear Regression

In [25]:
from statsmodels.formula.api import ols

In [26]:
import statsmodels.api as sm

C:\Users\Ryan May\Anaconda3\lib\site-packages\statsmodels\compat\pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


In [27]:
mq=ols('sale_price_clean~gross_square_feet+year', queens).fit()
print(mq.summary())

                            OLS Regression Results                            
Dep. Variable:       sale_price_clean   R-squared:                       0.053
Model:                            OLS   Adj. R-squared:                  0.053
Method:                 Least Squares   F-statistic:                     821.5
Date:                Tue, 29 May 2018   Prob (F-statistic):               0.00
Time:                        16:44:09   Log-Likelihood:            -4.7314e+05
No. Observations:               29099   AIC:                         9.463e+05
Df Residuals:                   29096   BIC:                         9.463e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
Intercept         -8.735e+07   5.18e+0

Above is the linear regression table for the sale price and the gross square feet of sales in Queens.  The coefficent of 17.1496 means that for each additional square foot generates and additional 17.1496 dollars in value on the property.  The year of the sale is interestly the most important variable at an additional $43,880 if a building is sold in 2016 rather than 2003.

In [28]:
mb=ols('sale_price_clean~gross_square_feet+year', brooklyn).fit()
print(mb.summary())

                            OLS Regression Results                            
Dep. Variable:       sale_price_clean   R-squared:                       0.167
Model:                            OLS   Adj. R-squared:                  0.167
Method:                 Least Squares   F-statistic:                     2357.
Date:                Tue, 29 May 2018   Prob (F-statistic):               0.00
Time:                        16:44:09   Log-Likelihood:            -3.8534e+05
No. Observations:               23431   AIC:                         7.707e+05
Df Residuals:                   23428   BIC:                         7.707e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
Intercept         -1.647e+08   6.95e+0

The coeffiencient for Brooklyn is quite higher at 99.4183, meaning that an additional square foot in Brooklyn is worth 5.79 times more than an additional square foot in Queens.  The year is even more important for Brooklyn at an additional $82,240 if a bulilding is sold in 2016 rather than 2003.

In [29]:
from sklearn.linear_model import LinearRegression

In [30]:
lm=LinearRegression()

In [31]:
queens.target=queens['sale_price_clean']

C:\Users\Ryan May\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  """Entry point for launching an IPython kernel.


In [32]:
X=pd.DataFrame(pd.get_dummies(queens[['building_class_category', 'neighborhood', 'tax_class_at_present', 'year', 'month', ]]))

In [33]:
lm.fit(X,queens.target)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [34]:
print('Estimated Intercept Coefficient {}' .format(lm.intercept_))

Estimated Intercept Coefficient 1.4611850903754476e+16


In [35]:
print('Estimated Coefficents{}' .format(lm.coef_), )

Estimated Coefficents[ 3.59233862e+04  9.79441731e+03  1.83047268e+16  1.83047268e+16
  1.83047268e+16  1.83047268e+16  1.83047268e+16  1.83047268e+16
  1.83047268e+16  1.83047268e+16  1.83047268e+16  1.83047268e+16
  1.83047269e+16  1.83047269e+16  1.83047268e+16  1.83047268e+16
  1.83047268e+16  1.83047268e+16  1.83047268e+16  1.83047268e+16
  1.83047268e+16  1.83047268e+16  1.83047268e+16  1.83047268e+16
  1.83047268e+16  1.83047269e+16  1.83047268e+16  1.83047268e+16
  1.83047268e+16  1.83047268e+16  1.09163719e+16  1.09163719e+16
  1.09163719e+16  1.09163719e+16  1.09163719e+16  1.09163719e+16
  1.09163719e+16  1.09163719e+16  1.09163719e+16  1.09163719e+16
  1.09163719e+16  1.09163719e+16  1.09163719e+16  1.09163719e+16
  1.09163719e+16  1.09163719e+16  1.09163719e+16  1.09163719e+16
  1.09163719e+16  1.09163719e+16  1.09163719e+16  1.09163719e+16
  1.09163719e+16  1.09163719e+16  1.09163719e+16  1.09163719e+16
  1.09163719e+16  1.09163719e+16  1.09163719e+16  1.09163719e+16
  1.

In [36]:
lm_queens=pd.DataFrame({'features': X.columns, 'estimatedCoefficients': lm.coef_})[['features', 'estimatedCoefficients']]

In [37]:
print(lm_queens.sort_values)

<bound method DataFrame.sort_values of                                              features  estimatedCoefficients
0                                                year           3.592339e+04
1                                               month           9.794417e+03
2          building_class_category_ASYLUMS_AND_HOMES_           1.830473e+16
3          building_class_category_COMMERCIAL_GARAGES           1.830473e+16
4      building_class_category_COMMERCIAL_VACANT_LAND           1.830473e+16
5                    building_class_category_CONDOPS_           1.830473e+16
6   building_class_category_COOPS_-_ELEVATOR_APART...           1.830473e+16
7   building_class_category_COOPS_-_WALKUP_APARTME...           1.830473e+16
8      building_class_category_EDUCATIONAL_FACILITIES           1.830473e+16
9                  building_class_category_FACTORIES_           1.830473e+16
10  building_class_category_HOSPITAL_AND_HEALTH_FA...           1.830473e+16
11  building_class_category_INDOOR_PU

In [38]:
brooklyn.target=brooklyn['sale_price_clean']

C:\Users\Ryan May\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  """Entry point for launching an IPython kernel.


In [39]:
Xb=pd.DataFrame(pd.get_dummies(brooklyn[['building_class_category', 'neighborhood', 'tax_class_at_present', 'year', 'month', ]]))

In [40]:
lm.fit(Xb,brooklyn.target)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [41]:
print('Estimated Intercept Coefficient {}' .format(lm.intercept_))

Estimated Intercept Coefficient 3148190189371888.5


In [42]:
print('Estimated Coefficents{}' .format(lm.coef_))

Estimated Coefficents[ 6.76297706e+04  4.98477781e+03 -1.16519722e+15 -1.16519717e+15
 -1.16519723e+15 -1.16519723e+15 -1.16519724e+15 -1.16519724e+15
 -1.16519723e+15 -1.16519723e+15 -1.16519723e+15 -1.16519723e+15
 -1.16519723e+15 -1.16519723e+15 -1.16519723e+15 -1.16519721e+15
 -1.16519723e+15 -1.16519723e+15 -1.16519723e+15 -1.16519722e+15
 -1.16519723e+15 -1.16519723e+15 -1.16519724e+15 -1.16519723e+15
 -1.16519723e+15 -1.16519723e+15 -1.16519721e+15 -1.16519723e+15
 -1.16519723e+15 -1.16519723e+15 -7.94066704e+13 -7.94066704e+13
 -7.94066705e+13 -7.94066705e+13 -7.94066703e+13 -7.94066695e+13
 -7.94066704e+13 -7.94066708e+13 -7.94066639e+13 -7.94066708e+13
 -7.94066706e+13 -7.94066708e+13 -7.94066707e+13 -7.94066696e+13
 -7.94066698e+13 -7.94066692e+13 -7.94066703e+13 -7.94066706e+13
 -7.94066705e+13 -7.94066708e+13 -7.94066633e+13 -7.94066589e+13
 -7.94066628e+13 -7.94066704e+13 -7.94066708e+13 -7.94066708e+13
 -7.94066708e+13 -7.94066713e+13 -7.94066706e+13 -7.94066708e+13
 -7.

In [43]:
lm_brooklyn=pd.DataFrame({'features': X.columns, 'estimatedCoefficients': lm.coef_})[['features', 'estimatedCoefficients']]

In [44]:
print(lm_brooklyn.sort_values)

<bound method DataFrame.sort_values of                                              features  estimatedCoefficients
0                                                year           6.762977e+04
1                                               month           4.984778e+03
2          building_class_category_ASYLUMS_AND_HOMES_          -1.165197e+15
3          building_class_category_COMMERCIAL_GARAGES          -1.165197e+15
4      building_class_category_COMMERCIAL_VACANT_LAND          -1.165197e+15
5                    building_class_category_CONDOPS_          -1.165197e+15
6   building_class_category_COOPS_-_ELEVATOR_APART...          -1.165197e+15
7   building_class_category_COOPS_-_WALKUP_APARTME...          -1.165197e+15
8      building_class_category_EDUCATIONAL_FACILITIES          -1.165197e+15
9                  building_class_category_FACTORIES_          -1.165197e+15
10  building_class_category_HOSPITAL_AND_HEALTH_FA...          -1.165197e+15
11  building_class_category_INDOOR_PU